[IAP] Error screens for upgrades flow

iamgabrielma opened the pull request:
## Description

This PR deals with the different error state views and their navigation flow (if needed). There's 5 potential errors we're dealing with:

- Error when fetching plans
- Error when loading merchant entitlements
- Error when attempting to upgrade more than one site 
- Purchase error
- In-App Purchases not supported

## Testing instructions
The easiest way to check all error screens is to force the specific case as `default` within the `UpgradesView`, rather than switching through all cases, and leave the `.loading` case as well just to see the state switch when we retry switching between states:

```
switch upgradesViewModel.upgradeViewState {
  case .loading:
    OwnerUpgradesView(upgradePlan: .skeletonPlan(), purchasePlanAction: {}, isLoading: true)
  default:
    let upgradeError = UpgradesError.purchaseError
    UpgradesErrorView(upgradeError,
      onRetryButtonTapped: {},
      onRetryPaymentButtonTapped: {},
      onCancelUpgradeTapped: {})
.padding(Layout.padding)
}
```

Then just switch through the different errors and load the preview or run the app for testing the button callback:

1. The cases for `.fetchError`, and `.entitlementsError` will show a `Retry` button. Tapping the button will attempt to fetch plans again:
```
let upgradeError = UpgradesError.fetchError
let upgradeError = UpgradesError.entitlementsError
```

2.  The case `.purchaseError` will show a `Try Payment Again` button which will retry fetching plans again and offer the option to tap on "Purchase" if a merchant wants to upgrade (same behaviour as the errors above), and a `Cancel Upgrade` button, which dismisses the view and redirects the merchant back to My Store:
```
let upgradeError = UpgradesError.purchaseError
```
3. Will show the screens above, without any button to interact with.
```
let upgradeError = UpgradesError.inAppPurchasesNotSupported
let upgradeError = UpgradesError.maximumSitesUpgraded
```

## Screenshots
| `fetchError`, `entitlementsError` | `maximumSitesUpgraded`  | `purchaseError` | `inAppPurchasesNotSupported`|
|--------|--------|--------|--------|
| <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/7969e325-3ade-407a-8421-a5a2087960a9"> | <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/e876216f-2e3d-4c0f-ae87-c5d749d3cc3f"> | <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/622fb470-afff-4bde-a386-b0e3a883ef6c"> | <img width=400 src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/ee2d14dc-04f9-49fd-8887-766b6c7c764f"> |

iamgabrielma committed:
Create UpgradesErrorView. Add UIImage for error screen.

iamgabrielma committed:
UI for `.fetchError` case. Add retry button

iamgabrielma committed:
UI for `.maximumSitesUpgraded` case

iamgabrielma committed:
Add `.entitlementsError` case

iamgabrielma committed:
UI for `.purchaseError` case, retry, & cancel buttons

iamgabrielma committed:
Add callbacks for error buttons

iamgabrielma committed:
Render error cases and style as per designs

iamgabrielma committed:
Center align text for error messages

iamgabrielma committed:
Update UI style for `.purchaseError` case

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -33,8 +33,7 @@ struct UpgradesView: View {
             CurrentPlanDetailsView(planName: subscriptionsViewModel.planName, daysLeft: subscriptionsViewModel.planDaysLeft)
-
-            Spacer()
+                .background(Color(UIColor.systemBackground))
```
Not urgent or required, just a note. Where possible, it's better to avoid the UIColor use... they don't always seem to get rendered correctly in previews, and it's just a bit clumsy. Specifying .systemBackground as a Color (in WooFoundation) would be a good thing to do some time after we ship M1 (or sooner if it's quick, but it's not at all urgent.)

iamgabrielma replied:
Sounds like a good improvement, I've logged this on a different issue here: https://github.com/woocommerce/woocommerce-ios/issues/10014

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -49,15 +48,143 @@ struct UpgradesView: View {
         }
         .navigationBarTitle(UpgradesView.Localization.navigationTitle)
         .padding(.top)
+        .background(Color(.listBackground).ignoresSafeArea())
```
As above for UIColor use

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -49,15 +48,143 @@ struct UpgradesView: View {
+
+    var body: some View {
+        VStack(alignment: .center) {
+            Image(uiImage: .planUpgradeError)
```
There's no benefit making a UIImage here – just use the Image with the filename, again it works better in previews and has less indirection. If you do it that way and the file doesn't exist, it'll show a placeholder instead, so there's no optionality to deal with.

iamgabrielma replied:
Gotcha! Changed here: 73f03156357dcb8b3a65b57ec83e4f83f473029c

joshheald commented on WooCommerce/Resources/Images.xcassets/plan-upgrade-error.imageset/plan-upgrade-error.svg:
This should probably be 3 PNGs, one for each size. While SVGs are great in theory, when they have raster graphics included they tend to become massive files.

iamgabrielma replied:
Makes sense, changed the SVG for PNG x3 here: 738711f0979185acbfd4166b3f0d09a54bd8c6ec

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -49,15 +48,143 @@ struct UpgradesView: View {
+            }
+        }
+        .padding(.bottom)
+        .background(Color(UIColor.systemBackground))
```
As above for UIColor

joshheald commented on WooCommerce/Classes/ViewRelated/Upgrades/UpgradesView.swift:
```diff
@@ -49,15 +48,143 @@ struct UpgradesView: View {
+                    Button(Localization.retryPaymentButtonText) {
+                        // TODO: Handle retry
+                        onRetryPaymentButtonTapped()
+                    }
```
This would be clearer if you move the display strings to properties on the error enum, possibly as a private extension of it in this file. Some of them (e.g. subtitle) may be optional and then you don't need to render that label if it's not present. e.g.

```
extension UpgradesError {
    var localizedDisplayMessage: String {
        switch self {
        case .fetchError:
            return Localization.fetchErrorMessage
        case .maximumSitesUpgraded, .entitlementsError:
            return Localization.maximumSitesUpgradedErrorMessage
        default:
            return ""
        }
    }

    var localizedDisplayDescription: String? {
        switch self {
        case .fetchError:
            return nil
        case .maximumSitesUpgraded, .entitlementsError:
            return Localization.maximumSitesUpgradedErrorSubtitle
        default:
            return nil
        }
    }

    var buttonText: String? {
        switch self {
        case .fetchError:
            return Localization.retry
        default:
            return nil
        }
    }
}
```

Simplifying the button so that you just pass in one optional action handler when you create it would also make it a lot easier to read and maintain.

I've deliberately ignored `.purchaseError` because it's going to be a totally different style so we don't need to worry about the differences.

iamgabrielma replied:
Looks cleaner indeed, thanks for the recommendation! I've logged this change here: https://github.com/woocommerce/woocommerce-ios/issues/10014 , as I can tackle it together when I make the changes for the updated `.purchaseError` views.

joshheald reviewed (commented):
Some thoughts... take them or leave them until later, none are essential for this PR.

iamgabrielma committed:
Move `.entitlementsError` case

Since this error will be shown if we have a problem loading entitlements, we can group it with the .fetchError case

iamgabrielma committed:
Update non-owner strings

iamgabrielma committed:
UI for `.inAppPurchasesNotSupported` case

iamgabrielma committed:
Add callback for fetch plans retry

iamgabrielma committed:
Handle cancelUpgradeTapped callback

iamgabrielma committed:
Add case for “retry payment button” tapped

iamgabrielma committed:
Remove unnecesary duplicated closure for retry action

iamgabrielma committed:
Update localizedstrings

iamgabrielma committed:
Update asset from svg to png x3

iamgabrielma committed:
Remove unused variables. Update access control

iamgabrielma committed:
Use Image rather than UIImage

iamgabrielma commented on WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift:
```diff
@@ -58,6 +58,15 @@ final class UpgradesViewModel: ObservableObject {
+    /// Sync wrapper for `fetchViewData`, so can be called directly from where this
+    /// ViewModel is referenced, outside of the initializer
+    ///
+    public func retryFetch() {
+        Task {
+            await fetchViewData()
+        }
+    }
```
We have a bit of duplication here, since this is only a wrapper for calling `fetchViewData`, but it felt adequate to duplicate the function so I didn't had to add async logic or Tasks to the view, neither deal with the state, or change the access control for the current `fetchViewData`. That said, I guess we could do just that as well and remove an extra method 🤔

joshheald replied:
Seems fine, I was similarly unsure of the neatest way earlier and I don't think there's much in it. Keeping the view simple seems good though, because SwiftUI can get so messy if you're not careful.

joshheald committed:
Update ErrorView spacing and padding to designs

joshheald reviewed (approved):
Tests well

I updated some of the padding and spacing to match the designs more closely in 8e93edc, however the image has a problem – lots of empty(ish) space at the bottom which shouldn't be there and is throwing our layout off. I've left it with that, and asked @joe-keenan for a new asset, which is a better way to fix it than compensating with smaller padding etc.

Still needs the `.maximumSitesUpgraded` error bringing in to this style though – are you going to do that separately?
